scheduler/src/cook/mesos/util.clj

```diff
@@ -337,20 +345,45 @@
-;; seek-datoms provides a pointer into the raw datomic indices 
-;; that we can then seek through. We set the pointer to look
-;; through the avet index, with attribute :job/user, seek to 
-;; user and then seek to the entity id that *would* have been
-;; created at expanded start. 
-;; This works because the submission time and job/user field
-;; are set at the same time, in "real" time. This means that
-;; jobs submitted after `start` will have been created after
-;; expanded start
+;; This function is O(# jobs in that state in the time range)
+(defn get-jobs-by-user-and-state-and-submit-by-state
+  "Returns all jobs for a particular user in the specified state
+   and timeframe, without a custom executor."
+  [db user start end state-keyword limit bucket]
+  (timers/time! (timers/timer ["cook-mesos" "scheduler" (str "get-" (name state-keyword) "-jobs-by-user-by-state-" bucket "-duration")])
+                (let [;; Expand the time range so that clock skew between cook
+                      ;; and datomic doesn't cause us to miss jobs
+                      ;; 1 hour was picked because a skew larger than that would be
+                      ;; suspicious
+                      expanded-start (Date. (- (.getTime start)
+                                               (-> 1 t/hours t/in-millis)))
+                      expanded-end (Date. (+ (.getTime end)
+                                             (-> 1 t/hours t/in-millis)))
+                      entid-start (d/entid-at db :db.part/user expanded-start)
+                      entid-end (d/entid-at db :db.part/user expanded-end)
+                      job-state-entid (d/entid db :job/state)
+                      state-entid (d/entid db state-keyword)
+                      ^long start-ms (.getTime start)
+                      ^long end-ms (.getTime end)
+                      result (->> (d/seek-datoms db :avet :job/state state-entid entid-start)
+                                  (take-while #(and (< (:e %) entid-end)
+                                                    (= (:a %) job-state-entid)
+                                                    (= (:v %) state-entid)))
+                                  (map :e)
+                                  (map (partial d/entity db))
+                                  (filter #(= (job-ent->user %) user))
+                                  (filter #(let [^long submit-ms (.getTime (:job/submit-time %))]
+                                             (and (<= start-ms submit-ms)
+                                                  (< submit-ms end-ms))))
+                                  (filter #(false? (:job/custom-executor %))))]
+                  ;; If we have a limit, take it here (before doall, so that our metrics are robust)
+                  (if limit (doall (take limit result)) result))))
+
+;; This differs from get-active-jobs-by-user-and-state as it is also looking up based on task state.
```

Comment:
This doesn't capture the only reason it differs. You should also include the comment block you removed below about why this one is querying on user instead of state.

Reply:
Fixed